compiler/front/commands.nim

```diff
@@ -1412,6 +183,186 @@ proc processArgument*(pass: TCmdLinePass; p: OptParser;
+          cliEvtErrCmdMissing:
+        discard
+
+const
+  cliLogAllKinds = {low(CliEventKind) .. high(CliEventKind)}
+  cliEvtErrors   = {cliEvtErrInvalidCommand .. cliEvtErrNoCliParamsProvided}
+  cliEvtWarnings = {cliEvtWarnSwitchValDeprecatedNoop}
+  cliEvtHints    = {cliEvtHintPathAdded}
+
+static:
+  const unaccountedForEvtKinds =
+    cliLogAllKinds - cliEvtErrors - cliEvtWarnings - cliEvtHints
+  doAssert unaccountedForEvtKinds == {}, "Uncategorized event kinds: " &
+                                            $unaccountedForEvtKinds
+
+proc procSwitchResultToEvents*(conf: ConfigRef, pass: TCmdLinePass,
+                               origParseOptKey, origParseOptVal: string,
+                               r: ProcSwitchResult): seq[CliEvent] =
+  # Note: the order in which this generates events is the order in which
+  #       they're output, rearrange code carefully.
+  case r.kind
+  of procSwitchSuccess: discard
+  else:
+    result.add:
+      CliEvent(kind: cliEvtErrFlagProcessing,
+                pass: pass,
+                origParseOptKey: origParseOptKey,
+                origParseOptVal: origParseOptVal,
+                procResult: r,
+                srcCodeOrigin: instLoc())
+  case r.switch
+  of cmdSwitchNimblepath:
+    if conf.hasHint(rextPath) and r.processedNimblePath.didProcess:
+      for res in r.processedNimblePath.nimblePathResult.addedPaths:
+        result.add:
+          CliEvent(kind: cliEvtHintPathAdded, pathAdded: res.string)
+  else:
+    discard
+
+proc writeLog(conf: ConfigRef, msg: string, evt: CliEvent) {.inline.} =
+  conf.writeLog(msg, evt.srcCodeOrigin)
+
+proc logError*(conf: ConfigRef, evt: CliEvent) =
+  # TODO: consolidate log event rendering, between this and "reports", but with
+  #       less of the reports baggage
+  func allowedCliOptionsArgs(flg: range[cliFlagMsgFormat..cliFlagMsgFormat]
+        ): seq[string] =
+    case flg
+    of cliFlagMsgFormat: @["text", "sexp"]
+
+  let msg =
+    case evt.kind
+    of cliEvtErrInvalidCommand:
+      "Invalid command - $1" % evt.cmd
+    of cliEvtErrCmdMissing:
+      "Command missing"
+    of cliEvtErrUnexpectedRunOpt:
+      "'$1' cannot handle --run" % evt.cmd
+    of cliEvtErrCmdExpectedNoAdditionalArgs:
+      "$1 command does not support additional arguments: '$2'" %
+        [evt.cmd, evt.unexpectedArgs]
+    of cliEvtErrRunCmdFailed,
+        cliEvtErrGenDependFailed: # make a better message for gen depend
+      "execution of an external program '$1' failed with exit code '$2'" %
+        [evt.shellCmd, $evt.exitCode]
+    of cliEvtErrNoCliParamsProvided:
+      "no command-line parameters provided"
+    of cliEvtErrFlagArgForbidden:
+      "$1 expects no arguments, but '$2' found" %
+        [evt.givenFlg, evt.givenArg]
+    of cliEvtErrFlagArgExpectedFromList:
+      "expected value for switch '$1'. Expected one of $2, but got nothing" %
+        [evt.givenFlg, allowedCliOptionsArgs(evt.flag).join(", ")]
```

Comment:
If I'm reading the code correctly, this is going to cause a range defect for all flags except `cliFlagMsgFormat`.

Reply:
Yeah, not well documented, but that's the only thing that can raise that particular error, I should pick a lane general implementation useable by all or specific implementation tied to the single use case.

Reply:
Creating a dedicated range and putting that into the variant results in an NPE in the compiler, I'll trouble shoot that later/separately.